File: apps/agent/src/agents/workspaceAgent/projections/task.ts
import type { TaskProjection } from '@lifecoach/shared-types';

/**
 * Project a raw `tasks.tasks.list` / `tasks.tasks.get` response into the
 * shape the LLM consumes. Drops fields the coach doesn't need
 * (`etag`, `selfLink`, `kind`, `updated`, `position`, `links`).
 */

export interface RawTask {
  id?: string;
  title?: string;
  due?: string;
  status?: string;
  notes?: string;
  completed?: string;
}

export function projectTask(raw: RawTask, taskListId: string): TaskProjection {
  const status: TaskProjection['status'] = raw.status === 'completed' ? 'completed' : 'needsAction';

  const projection: TaskProjection = {
    id: raw.id ?? '',
    taskListId,
    title: raw.title ?? '(untitled)',
    status,
  };
  if (raw.due) projection.due = raw.due;
  if (raw.notes) projection.notes = raw.notes;
  if (raw.completed) projection.completed = raw.completed;

  return projection;
}

File: apps/agent_py/src/lifecoach_agent/workspace_agent/projections/task.py
"""Project a raw `tasks.tasks.list` / `tasks.tasks.get` response into the
shape the LLM consumes. Drops fields the coach doesn't need (`etag`,
`selfLink`, `kind`, `updated`, `position`, `links`).
"""

from __future__ import annotations

from typing import Any

from lifecoach_agent.contracts.models import TaskProjection


def project_task(raw: dict[str, Any], task_list_id: str) -> TaskProjection:
    status = "completed" if raw.get("status") == "completed" else "needsAction"
    payload: dict[str, Any] = {
        "id": raw.get("id") or "",
        "taskListId": task_list_id,
        "title": raw.get("title") or "(untitled)",
        "status": status,
    }
    if raw.get("due"):
        payload["due"] = raw["due"]
    if raw.get("notes"):
        payload["notes"] = raw["notes"]
    if raw.get("completed"):
        payload["completed"] = raw["completed"]
    return TaskProjection.model_validate(payload)
